### ISSUE TITLE
[QUERY] Identifying the requested scope when validating the JWT

### ISSUE THREAD

**Issue by martfol** [Azure/azure-sdk-for-java#38237]
**Query/Question**
Hi guys, I'm trying to implmement a scope based JWT validation, but I'm not able to identify the requested scope of the incoming token.  Here's my setup followed by the question....

I have two applications registered in Azure - A and B. Application B (=resource) exposes two scopes - scope1 and scope2. Application A (=client) has configured permissions to access both these scopes. This is how the _application.properties_ configuration looks like in the client application A:

```
spring.cloud.azure.active-directory.authorization-clients.scope1ID.scopes=api://B/scope1
spring.cloud.azure.active-directory.authorization-clients.scope1ID.authorizationGrantType=azure_delegated
spring.cloud.azure.active-directory.authorization-clients.scope2ID.scopes=api://B/scope2
spring.cloud.azure.active-directory.authorization-clients.scope2ID.authorizationGrantType=azure_delegated
```

The business logic within the client application A then requests an access token either for scope1 or scope2, e.g.:
```
(1) OAuth2AuthorizeRequest request = OAuth2AuthorizeRequest.withClientRegistrationId("scope1ID") // access token requested for scope1
(2)        .principal(authentication).build();
(3) Optional<OAuth2AccessToken> accessToken = Optional.empty();
(4) OAuth2AuthorizedClient authorize = authorizedClientManager.authorize(request);
(5) if (authorize != null) {
(6)      accessToken = Optional.of(authorize.getAccessToken());
(7) }
```

Then, application A sends request to application B. In application B I create my own `JwtDecoder `with several custom validators implementing `OAuth2TokenValidator<Jwt>`.

Now the problem is, that the JWT token - as mentioned [here](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-configure-app-expose-web-apis#using-the-exposed-scopes) - contains in its _scp_ claim **all the permissions** I've specified in the client's app registration, i.e. both "scope1" and "scope2".

However one of my custom validators requires to know, **which scope was requested/used when obtaining the access token** (see line (1) above)? Is it somehow possible to "insert" this information into the claims of the particular JWT, so I can evaluate it in my custom JWT validator in the application B? Or is it generally not possible and my business logic is wrong if I need this kind of information? Or is there a workaround or trick how to achieve this functionality? Any tips or tricks appreciated... Thanks a lot in advance!

***Why is this not a Bug or a feature Request?***
This not a bug or a feature request, it's a request for help :)

**Setup (please complete the following information if applicable):**
 - Azure SDK 4.11.0
 - spring boot 2.7.18
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by Netyyyy** [comment]
HI @martfol , thanks for reaching out.
For your properties, are the `scope1ID`, `scope2ID` your application names? and could help provide more information?
Here is a [sample](https://github.com/Azure-Samples/azure-spring-boot-samples/tree/main/aad/spring-cloud-azure-starter-active-directory/web-client-access-resource-server) that may help you.

**Comment by martfol** [comment]
Hi,

`scope1ID` and `scope2ID` are just "client registration IDs", i.e. if I understand it right, those are just "random strings" used as mappings to find the corresponding _spring.cloud.azure.active-directory.authorization-clients..._ properties.

I know realised I probably described the problem in a very complicated way. Let me rephrase my question:

How are we supposed to implement scope-based authorization in the resource server, when the JWT token allways contains **all the permissions** of the client in the "scp" claim, not jut the one used to create the request (i.e. the one being used by the client application to access the resource server)...?

**Comment by moarychan** [comment]
Hi @martfol , thanks for your update!
If the JWT token always contains all the scopes,  I am not sure why, but you can try to update your app A configuration and get a single scope-based token.

```
spring.cloud.azure.active-directory.authorization-clients.scope1ID.scopes=api://B/scope1
```

If still the same response, you may [raise an issue](https://github.com/MicrosoftDocs/azure-docs/issues) based on this doc [oauth2-auth-code-flow](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow).
